openshift/console
# Bug 1839621: Add an ability to proxy requests to Che Workspace

### What does this PR do?
It solves https://bugzilla.redhat.com/show_bug.cgi?id=1839621

So, it adds backend endpoints 

- `/api/terminal/proxy/[namespace]/[workspace-name]/[path]`: proxy requests on `[path]` to workspace in given namespace, attaching user's bearer token. This endpoint is used to construct a kubeconfig in the user's workspace, allowing them to access commandline tools.
- `/api/terminal/available/`: Returns `204` if proxy is available, `503` otherwise. Used to check if the endpoint above will process requests (see security considerations below)

This PR is needed for the OpenShift Console to open a terminal. See screencast how it works:
<details>
<summary>Screencast</summary>

![console-terminal](https://user-images.githubusercontent.com/5887312/81664853-9b2a0580-9448-11ea-9d02-4ea3015f3277.gif)

</details>

### Setup
**Note:** To ease testing, this PR includes some commits that include an early implementation of the frontend UI that will be used. These commits will be dropped from this PR before merging. The nature of the changes means it's much harder to test otherwise.


1. Deploy changes from this PR

Built image could be used:
```
oc patch consoles.operator.openshift.io cluster --patch '{ "spec": { "managementState": "Unmanaged" } }' --type=merge
oc set image deploy console console=sleshchenko/console:apiTerminal-27-05-2020 -n openshift-console
```
^ image has date in the tag for you to make sure it's up to date, and it will be updated after any changes in the PR. 


2. Set up che-workspace-controller
    ```
    git clone https://github.com/che-incubator/che-workspace-operator.git
    cd che-workspace-operator
    export WEBHOOK_ENABLED=true
    export DEFAULT_ROUTING=basic
    make deploy
    ```
    Changing `WEBHOOK_ENABLED` above to false should disable all endpoints

### Testing
Note: it's necessary to grant additional permissions to the `console` SA, as in https://github.com/openshift/console-operator/pull/432

#### Testing the `/api/terminal/available/` endpoint

use curl from terminal
```
curl -k -H "Authorization: Bearer $(oc whoami -t)" \
  ${cluster_URL}/api/terminal/available/ --verbose
```
Should return `204` if `WEBHOOK_ENABLED`, `503` otherwise.

#### Testing the main proxy endpoint:
Cases:
- Log in as a cluster-admin user, click terminal button and wait for workspace to be running. Should see a request sent to `.../exec/init` with 403 response, and message "Terminal is disabled for cluster-admin users."
![Screenshot from 2020-05-21 14-02-01](https://user-images.githubusercontent.com/16168279/82594337-ebecdd00-9b71-11ea-96ce-0a5dc45fa914.png)


- Log in as regular user, click terminal button. Should see a request to `.../exec/init` with 200 status, and terminal should start. Once terminal is initialized, `oc whoami` should list current user as logged in. **Note:** A few front-end workarounds are required to make this work:
  - Front end currently depends on listing workspaces at cluster scope. To enable this easily, bind the view-workspaces clusterrole to the user:
    ```
    oc adm policy add-cluster-role-to-user view-workspaces ${testuser}
    ```
  - On first run, the terminal seems to get stuck loading. It's necessary to close the panel and reopen it for the terminal to initialize.
  
  You can also exec directly into the `dev` container in the workspace pod, and test `oc` there.

In both cases, if `WEBHOOK_ENABLED=false` when deploying the operator, the requests to `.../exec/init` should return 403 with message "Terminal endpoint is disabled: workspace operator is not deployed."

### Workflow implementation details
The current implementation uses the following flow:

1. Frontend creates a namespace (if necessary) and initializes a Che workspace
2. The che-workspace-operator creates a deployment for the requested workspace
3. The frontend POSTs the user's token to the backend proxy, which forwards it to the workspace's /exec/init endpoint
4. The workspace component creates a kubeconfig file for the passed token
5. The front end does a regular exec into the workspace container

### Related issues/PRs
It depends on https://github.com/openshift/console-operator/pull/432

### Security concerns

We're adding a proxy to the backend that is submitting a user's auth token to an URL defined in a status field. As a result it's necessary to carefully consider cases to avoid sending the auth token to an unintended recipient. The current security model on the workspace operator side is:

- Upon creation, workspaces get a creator annotation (org.eclipse.che.workspace/creator) that is set to the user's UID. This is protected by a webhook, and the annotation is read on the backend to ensure the token is being sent to a workspace that was created by the current user.
- Exec access to any containers in a workspace is restricted to the creator, via webhook

To support these backend changes, requests to the new backend endpoints perform some checks before proxying the user's token:

- Check if user is cluster-admin or otherwise has strong privileges, and if so deny the request.
  - This is done by checking if the user token can create pods in the `openshift-operators` namespace
- Check that webhooks deployed by the che-workspace-operator are present on the cluster, and if not, deny the request.
